Source/JavaScriptCore/runtime/WaiterListManager.h

```diff
@@ -0,0 +1,226 @@
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#pragma once
+
+#include "DeferredWorkTimer.h"
+#include "JSPromise.h"
+#include <wtf/Condition.h>
+#include <wtf/HashMap.h>
+#include <wtf/Lock.h>
+#include <wtf/SentinelLinkedList.h>
+
+namespace JSC {
+
+enum class AtomicsWaitType : uint8_t { Sync, Async };
+enum class AtomicsWaitValidation : uint8_t { Pass, Fail };
+
+class Waiter final : public WTF::BasicRawSentinelNode<Waiter>, public ThreadSafeRefCounted<Waiter> {
+    WTF_MAKE_FAST_ALLOCATED;
+
+public:
+    Waiter(VM* vm)
+        : m_isAsync(false)
+    {
+        m_vm = vm;
+    }
+
+    Waiter(JSPromise* promise)
+        : m_isAsync(true)
+    {
+        VM& vm = promise->vm();
+        {
+            JSLockHolder locker(vm);
+            m_ticket = vm.deferredWorkTimer->addPendingWork(vm, promise, { });
+        }
+    }
+
+    bool isAsync()
+    {
+        return m_isAsync;
+    }
+
+    VM* getVM()
+    {
+        return m_isAsync ? &jsCast<JSPromise*>(m_ticket->target())->vm() : m_vm;
+    }
+
+    void nullOutVM()
+    {
+        ASSERT(!m_isAsync);
+        m_vm = nullptr;
+    }
+
+    Condition& getCondition()
+    {
+        ASSERT(!m_isAsync);
+        return m_condition;
+    }
+
+    DeferredWorkTimer::Ticket getTicket()
+    {
+        ASSERT(m_isAsync);
+        return m_ticket;
+    }
+
+    void setTimer(Ref<RunLoop::DispatchTimer>&& timer)
+    {
+        ASSERT(m_isAsync);
+        m_timer = WTFMove(timer);
+    }
+
+    bool hasTimer()
+    {
+        return !!m_timer;
+    }
+
+    void cancelTimer()
+    {
+        ASSERT(m_isAsync);
+        // If the timeout for AsyncWaiter is infinity, we won't dispatch any timer.
+        if (!m_timer)
+            return;
+        m_timer->stop();
+        m_timer->deref();
```

Comment:
This is a memory corruption bug. `RefPtr` does automatic `ref` / `deref`. If you call it manually, it broke ref count and corrupting memory.
Let's make it `m_timer = nullptr;`.